File: frontend/src/api/network.ts
import { api } from './request'
import type {
  NetworkInterface, PortInfo, TaskResponse,
  FirewallStatus, BondsResponse, BondOptionDef,
  VlanInfo, BridgeInfo, RouteInfo, WireGuardInfo, WireGuardCheckResult,
} from '@/types/api'

export const networkApi = {
  getInterfaces: () => api<NetworkInterface[]>('/network/interfaces'),
  getDns: () => api<{ dns: string[] }>('/network/dns'),
  setDns: (servers: string[]) => api<{ success: boolean; message: string }>('/network/dns', { method: 'POST', body: { servers } }),
  getFirewall: () => api<FirewallStatus>('/network/firewall'),
  getFirewallRules: (tool: string) => api<{ rules: string }>(`/network/firewall/rules?tool=${tool}`),
  getNetworkManager: () => api<{ manager: string }>('/network/manager'),
  // @deprecated — use interfaceAction / fwAllow / fwDeny for structured operations
  runFwCmd: (cmd: string) => api<{ success: boolean; message: string }>('/network/fwcmd', { method: 'POST', body: { cmd } }),
  getPorts: () => api<{ ports: PortInfo[] }>('/network/ports'),
  // Semantic endpoints
  interfaceAction: (name: string, action: 'up' | 'down' | 'restart') =>
    api<{ success: boolean; message: string }>('/network/interface/action', { method: 'POST', body: { name, action } }),
  fwAllow: (port: string, protocol: string) =>
    api<{ success: boolean; message: string }>('/network/firewall/allow', { method: 'POST', body: { port, protocol } }),
  fwDeny: (port: string, protocol: string) =>
    api<{ success: boolean; message: string }>('/network/firewall/deny', { method: 'POST', body: { port, protocol } }),
  // IP config
  setStaticIp: (iface: string, address: string, netmask?: string, gateway?: string, dns?: string) =>
    api<{ success: boolean; message: string }>('/network/ip/static', { method: 'POST', body: { interface: iface, address, netmask, gateway, dns } }),
  setDhcp: (iface: string) => api<{ success: boolean; message: string }>('/network/ip/dhcp', { method: 'POST', body: { interface: iface } }),
  getIfaceIpMode: (iface: string) => api<{ mode: string; netmask: string; gateway: string; dns: string }>(`/network/ip-config/${iface}`),
  getTraffic: () => api<{ interfaces: any[]; total_rx: number; total_tx: number }>('/network/traffic'),
  getFirewallZones: () => api<any>('/network/firewall/zones'),
  getFirewallRichRules: (zone?: string) => api<{ rules: any[] }>(`/network/firewall/rich-rules${zone ? `?zone=${zone}` : ''}`),
  // Diagnostics
  diagPing: (host: string, count?: number) => api<any>(`/network/diag/ping?host=${encodeURIComponent(host)}&count=${count||4}`),
  diagTraceroute: (host: string) => api<any>(`/network/diag/traceroute?host=${encodeURIComponent(host)}`),
  diagDns: (domain: string, type?: string) => api<any>(`/network/diag/dns?domain=${encodeURIComponent(domain)}&type=${type||'A'}`),
  diagPortscan: (host: string, ports?: string) => api<any>(`/network/diag/portscan?host=${encodeURIComponent(host)}&ports=${ports||''}`),
  diagConnectivity: (target?: string) => api<any>(`/network/diag/connectivity${target ? `?target=${target}` : ''}`),

  // Bonding
  getBonds: () => api<BondsResponse>('/network/bonds'),
  createBond: (name: string, slaves: string[], mode?: string, ip?: string, gateway?: string) =>
    api<{ success: boolean; message: string }>('/network/bond/create', { method: 'POST', body: { name, slaves, mode, ip, gateway } }),
  deleteBond: (name: string) => api<{ success: boolean; message: string }>('/network/bond/delete', { method: 'POST', body: { name } }),
  getBondOptions: () => api<{ options: Record<string, BondOptionDef> }>('/network/bond/options'),
  createBondAdvanced: (name: string, slaves: string[], mode?: string, options?: Record<string, string>) =>
    api<{ success: boolean; message: string }>('/network/bond/create-advanced', { method: 'POST', body: { name, slaves, mode, options } }),
  // VLAN
  getVlans: () => api<{ vlans: VlanInfo[] }>('/network/vlans'),
  createVlan: (parent: string, vlan_id: number, name?: string) =>
    api<{ success: boolean; message: string }>('/network/vlan/create', { method: 'POST', body: { parent, vlan_id, name } }),
  deleteVlan: (name: string) => api<{ success: boolean; message: string }>('/network/vlan/delete', { method: 'POST', body: { name } }),
  // Bridge
  getBridges: () => api<{ bridges: BridgeInfo[] }>('/network/bridges'),
  createBridge: (name: string) => api<{ success: boolean; message: string }>('/network/bridge/create', { method: 'POST', body: { name } }),
  bridgeAddMember: (bridge: string, iface: string) =>
    api<{ success: boolean; message: string }>('/network/bridge/add-member', { method: 'POST', body: { bridge, interface: iface } }),
  bridgeRemoveMember: (iface: string) =>
    api<{ success: boolean; message: string }>('/network/bridge/remove-member', { method: 'POST', body: { interface: iface } }),
  deleteBridge: (name: string) => api<{ success: boolean; message: string }>('/network/bridge/delete', { method: 'POST', body: { name } }),
  // Routes
  getRoutes: () => api<{ routes: RouteInfo[] }>('/network/routes'),
  addRoute: (dst: string, gateway?: string, dev?: string, metric?: string) =>
    api<{ success: boolean; message: string }>('/network/route/add', { method: 'POST', body: { dst, gateway, dev, metric } }),
  deleteRoute: (dst: string, gateway?: string) =>
    api<{ success: boolean; message: string }>('/network/route/delete', { method: 'POST', body: { dst, gateway } }),
  // WireGuard
  checkWireguard: () => api<WireGuardCheckResult>('/network/wireguard/check'),
  getWireguardList: () => api<{ interfaces: WireGuardInfo[] }>('/network/wireguard/list'),
  createWireguard: (name: string) => api<{ success: boolean; message: string }>('/network/wireguard/create', { method: 'POST', body: { name } }),
  deleteWireguard: (name: string) => api<{ success: boolean; message: string }>('/network/wireguard/delete', { method: 'POST', body: { name } }),
}

File: backend/routes/network.py
"""Network routes."""
import re
from flask import Blueprint, jsonify
from utils.helpers import safe_api, validate_json, require_auth, validate_fw_command, run_cmd, safe_quote
from core import network

bp = Blueprint("network", __name__)


@bp.route("/api/network/interfaces")
@safe_api
@require_auth
def interfaces():
    return jsonify(network.get_network_interfaces())


@bp.route("/api/network/dns")
@safe_api
@require_auth
def dns():
    return jsonify({"dns": network.get_dns()})


@bp.route("/api/network/firewall")
@safe_api
@require_auth
def firewall():
    return jsonify(network.get_firewall_status())


@bp.route("/api/network/fwcmd", methods=["POST"])
@safe_api
@require_auth
@validate_json(["cmd"])
def fwcmd(data):
    ok, validated = validate_fw_command(data["cmd"])
    if not ok:
        return jsonify({"success": False, "message": validated}), 403
    out, code = run_cmd(f"sudo {validated}", timeout=15)
    return jsonify({"success": code == 0, "message": out})


@bp.route("/api/network/dns", methods=["POST"])
@safe_api
@require_auth
@validate_json(["servers"])
def set_dns(data):
    return jsonify(network.set_dns(data["servers"]))


@bp.route("/api/network/ports")
@safe_api
@require_auth
def ports():
    return jsonify({"ports": network.get_listen_ports()})


# ── Semantic endpoints (replacing frontend shell-command construction) ──

@bp.route("/api/network/interface/action", methods=["POST"])
@safe_api
@require_auth
@validate_json(["name", "action"])
def interface_action(data):
    """Bring an interface up, down, or restart it."""
    name = data["name"].strip()
    action = data["action"].strip().lower()
    if not re.match(r'^[a-zA-Z0-9._\-:]+$', name):
        return jsonify({"success": False, "message": "Invalid interface name"}), 400
    if action not in ("up", "down", "restart"):
        return jsonify({"success": False, "message": "Action must be up, down, or restart"}), 400
    if action == "restart":
        out, code = run_cmd(f"sudo ip link set {safe_quote(name)} down && sudo ip link set {safe_quote(name)} up", timeout=15)
    else:
        out, code = run_cmd(f"sudo ip link set {safe_quote(name)} {action}", timeout=10)
    return jsonify({"success": code == 0, "message": out or f"Interface {name} {action}"})


@bp.route("/api/network/firewall/allow", methods=["POST"])
@safe_api
@require_auth
@validate_json(["port"])
def fw_allow(data):
    """Allow a port through the firewall."""
    port = str(data["port"]).strip()
    protocol = data.get("protocol", "tcp").strip().lower()
    if not port.isdigit() or not (1 <= int(port) <= 65535):
        return jsonify({"success": False, "message": "Invalid port number (1-65535)"}), 400
    if protocol not in ("tcp", "udp"):
        return jsonify({"success": False, "message": "Protocol must be tcp or udp"}), 400
    out, code = run_cmd(f"sudo ufw allow {port}/{protocol}", timeout=15)
    return jsonify({"success": code == 0, "message": out})


@bp.route("/api/network/firewall/deny", methods=["POST"])
@safe_api
@require_auth
@validate_json(["port"])
def fw_deny(data):
    """Deny a port through the firewall."""
    port = str(data["port"]).strip()
    protocol = data.get("protocol", "tcp").strip().lower()
    if not port.isdigit() or not (1 <= int(port) <= 65535):
        return jsonify({"success": False, "message": "Invalid port number (1-65535)"}), 400
    if protocol not in ("tcp", "udp"):
        return jsonify({"success": False, "message": "Protocol must be tcp or udp"}), 400
    out, code = run_cmd(f"sudo ufw deny {port}/{protocol}", timeout=15)
    return jsonify({"success": code == 0, "message": out})
